### PR TITLE
Final DATAPTR and USE_RINTERNALS removal

### LINKED ISSUE #3301: Remove #define USE_RINTERNALS
With ALTREP in R 3.5+ it's no longer true that defining USE_RINTERNALS before including Rinternals.h makes REAL(), INTEGER() etc into macros.   That's why I did it originally and it's no longer true, so there's no point in doing it anymore. It was good advice because it was a lot faster. The new advice with R 3.5+ is to not use the R API inside loops.
See also: https://github.com/Rdatatable/data.table/issues/2866#issuecomment-389643625 and https://github.com/Rdatatable/data.table/issues/3188#issuecomment-444757748

Section 5.14 of R-exts manual contains : 

> If efficiency is essential, the internal versions of the accessors can be obtained by defining ‘USE_RINTERNALS’ before including Rinternals.h. If you find it necessary to do so, please do test that your code compiles without ‘USE_RINTERNALS’ defined, as this provides a stricter test that the accessors have been used correctly. Also be prepared to adjust your code should R internals change. 
> The accessor functions, and other functions in the R API, are also subject to change to support the ‘ALTREP’ project (https://svn.r-project.org/R/branches/ALTREP/ALTREP.html). Code that does not define ‘USE_RINTERNALS’ should not be affected by these changes, but code that does define ‘USE_RINTERNALS’ may need to be adjusted.

I've tried turning off USE_RINTERNALS and data.table compiles and passes all tests with latest versions of R because DATAPTR is exposed in the R API. But still we're not supposed to use it. Not passing R 3.1.0 until DATAPTR is removed (~72~ 22 calls) : 
- [x] assign.c (#3610 #4095; v1.12.8)
- [x] bmerge.c (#3610; v1.12.4)
- [x] dogroups.c (#3610 #4164; v1.12.10)
- [x] fmelt.c (#3350; v1.12.2)
- [x] freadR.c [#4164; v1.12.10]
- [x] fwriteR.c [#4164; v1.12.10]
- [x] gsumm.c (#3350; v1.12.2)
- [x] quickselect.c (#3377; v1.12.2)
- [x] rbindlist.c (#3350 #3455; v1.12.2)
- [x] reorder.c (#4164; v1.12.10)
- [x] shift.c (#3348; v1.12.2)
- [x] uniqlist.c (#3350; v1.12.2)

### PR DESCRIPTION
**Pull request by mattdowle**
Closes #3301

### CODE REVIEW

File: src/utils.c [hunk 1]
```diff
@@ -355,3 +355,15 @@ SEXP coerceUtf8IfNeeded(SEXP x) {
   UNPROTECT(1);
   return(ans);
 }
+
+//const void *DATAPTR_RO(SEXP x) {
```
Comment (jangorecki) [inline src/utils.c]: good catch about backporting that...
```
/builds/jangorecki/data.table/bus/test-310-cran-lin/data.table.Rcheck/data.table/libs/datatable.so: undefined symbol: DATAPTR_RO
```

Comment (MichaelChirico) [reply src/utils.c]: IIUC was added in roughly Sep '17 ~ 3.4

https://github.com/wch/r-source/commit/447cdb613a16c6419b6b9cc6bcfa9d9554f97db2

**Comment by mattdowle** [comment]
The travis fail appears to be due to `stringi` updated on CRAN today.  Where does `stringi` come into the CI dependency chain?  It's not in the data.table repo afaics.

**Comment by jangorecki** [comment]
It is required for knitr via stringr

**Comment by MichaelChirico** [comment]
What's devtools used for? github install?

`remotes` used for the same and has lighter footprint

File: src/reorder.c [hunk 2]
```diff
@@ -37,66 +37,61 @@ SEXP reorder(SEXP x, SEXP order)
...
+    // when we're sure we're passing in a strict permutation, and ii) this operation is fundamental to setkey and data.table.
   }
-  // Creorder is for internal use (so we should get the input right!), but the check above seems sensible. The for loop above should run
-  // in neglible time (sequential with prefetch). It will catch NAs anywhere but won't catch duplicates. But doing so would be going too
-  // far given this is for internal use only and we only use this function internally when we're sure it's a permutation.
 
-  char *TMP = malloc(nrow * maxSize);
-  // enough RAM for a copy of one column (of largest type). Writes into the [start,end] subset. Outside [start,end] is wasted in that rarer case
-  // to save a "-start" in the deep loop below in all cases.
+  char *TMP = malloc((end-start+1)*maxSize);
```
Comment (jangorecki) [inline src/reorder.c]: Any reason why we don't use R_alloc here?

Comment (mattdowle) [reply src/reorder.c]: Nope. Good spot.

File: src/utils.c (hunk 1 above)
Comment (mattdowle) [reply src/utils.c]: To double-check, I grabbed the released tar.gz for R 3.4.4 and 3.5.0.  DATAPTR_RO was new in 3.5.0.  It's not always clear from the R-devel commits what is 'backported' to the next minor release and what is intended for the next major release.  So although it was committed during the R 3.4 series,  it wasn't released until R 3.5.0.

**Comment by mattdowle** [comment]
We can do review via follow up PRs.  Merging now to make progress, see if it passes stronger GLCI, and to push to anyone testing dev.

**Comment by mattdowle** [comment]
Good. Passing GLCI (e.g. R 3.1.0).
I submitted to win-builder both R-devel and R-devel-ATC and both passed fully ok, too.
@jangorecki Since R 3.5.0 was a significant change point for data.table, how time consuming would it be to add two more tests to GLCI: R 3.4.4 and R 3.5.0.  To test just before and just after ALTREP & DATAPTR_RO. That way we test that we've got the version switch branches correct.

File: src/frank.c [hunk 3]
```diff
@@ -81,42 +78,43 @@ SEXP frank(SEXP xorderArg, SEXP xstartArg, SEXP xlenArg, SEXP ties_method) {
...
   // else if (!strcmp(CHAR(STRING_ELT(ties_method, 0)), "runlength")) ties = RUNLENGTH;
   else error(_("Internal error: invalid ties.method for frankv(), should have been caught before. please report to data.table issue tracker")); // # nocov
-  n = length(xorderArg);
-  SEXP ans = (ties == MEAN) ? PROTECT(allocVector(REALSXP, n)) : PROTECT(allocVector(INTSXP, n));
-  int *ians = INTEGER(ans);
-  double *dans = REAL(ans);
-  if (n > 0) {
+  const int n = length(xorderArg);
+  SEXP ans = PROTECT(allocVector(ties==MEAN ? REALSXP : INTSXP, n));
+  int *ians=NULL;
```
Comment (MichaelChirico) [inline src/frank.c]: why initialize to `NULL`, can't we just do `int *ians;` and `double *dans;`?

File: src/nafill.c [hunk 4]
```diff
@@ -121,16 +121,26 @@ SEXP nafillR(SEXP obj, SEXP type, SEXP fill, SEXP nan_is_na_arg, SEXP inplace, S
...
+    }
   }
   if (!binplace) {
     ans = PROTECT(allocVector(VECSXP, nx)); protecti++;
     for (R_len_t i=0; i<nx; i++) {
       SET_VECTOR_ELT(ans, i, allocVector(TYPEOF(VECTOR_ELT(x, i)), inx[i]));
-      vans[i] = ((ans_t) { .dbl_v=REAL(VECTOR_ELT(ans, i)), .int_v=INTEGER(VECTOR_ELT(ans, i)), .int64_v=(int64_t *)REAL(VECTOR_ELT(ans, i)), .status=0, .message={"\0","\0","\0","\0"} });
+      const SEXP ansi = VECTOR_ELT(ans, i);
+      const void *p = isReal(ansi) ? (void *)REAL(ansi) : (void *)INTEGER(ansi);
+      vans[i] = ((ans_t) { .dbl_v=(double *)p, .int_v=(int *)p, .int64_v=(int64_t *)p, .status=0, .message={"\0","\0","\0","\0"} });
```
Comment (MichaelChirico) [inline src/nafill.c]: @jangorecki doesn't `""` null-terminate itself? Or are we double-null-terminating these?

**Comment by MichaelChirico** [comment]
Gave a read and it looks good to me

File: src/frank.c (hunk 3 above)
Comment (jangorecki) [reply src/frank.c]: AFAIU to avoid compiler warnings about uninitialized variables

File: src/nafill.c (hunk 4 above)
Comment (jangorecki) [reply src/nafill.c]: No idea

Comment (mattdowle) [reply src/nafill.c]: Looks like those are double-null terminated. I don't know why either.

File: src/nafill.c [hunk 5]
```diff
@@ -121,16 +121,26 @@ SEXP nafillR(SEXP obj, SEXP type, SEXP fill, SEXP nan_is_na_arg, SEXP inplace, S
   SEXP ans = R_NilValue;
   ans_t *vans = (ans_t *)R_alloc(nx, sizeof(ans_t));
   for (R_len_t i=0; i<nx; i++) {
-    inx[i] = xlength(VECTOR_ELT(x, i));
-    dx[i] = REAL(VECTOR_ELT(x, i));
-    ix[i] = INTEGER(VECTOR_ELT(x, i));
-    i64x[i] = (int64_t *)REAL(VECTOR_ELT(x, i));
+    const SEXP xi = VECTOR_ELT(x, i);
+    inx[i] = xlength(xi);
+    // not sure why these pointers are being constructed like this; TODO: simplify structure
```
[the remark this answers is not in the document]
Comment (mattdowle) [reply src/nafill.c]: I'll try and come back and suggest something. There must be a simpler way.

File: src/dogroups.c [hunk 6]
```diff
@@ -113,150 +114,74 @@ SEXP dogroups(SEXP dt, SEXP dtcols, SEXP groups, SEXP grpcols, SEXP jiscols, SEX
...
-            // #3634 -- CPLXSXP columns have size 16
-            Rcomplex *td = COMPLEX(target);
-            const Rcomplex *sd = COMPLEX(source);
-            for (int k=0; k<grpn; ++k) {
-              rownum = iI[k]-1;
-              td[k] = sd[rownum];
-            }
-          }
+          // this is the main non-contiguous gather, and is parallel (within-column) for non-SEXP
+          subsetVectorRaw(VECTOR_ELT(SDall,j), VECTOR_ELT(dt,INTEGER(dtcols)[j]-1), I, /*anyNA=*/false);
```
[the remark this answers is not in the document]
Comment (jangorecki) [reply src/dogroups.c:175]: seems that `_Pragma("...")` that we use inside `subsetVectorRaw` was giving an overhead also for 1 threaded computation, fix for that is in https://github.com/Rdatatable/data.table/pull/4558

**Pull request merged**